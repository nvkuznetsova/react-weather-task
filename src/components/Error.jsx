import React from 'react';

export const Error = (props) => {
  return (
    <div className="error">{props.message}</div>
  );
}