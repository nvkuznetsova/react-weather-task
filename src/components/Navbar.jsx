import React from 'react';
import { NavLink } from 'react-router-dom';

export const Navbar = () => {
  const moscowId = 524901;
  const rostovId = 501175;
  return (
    <nav className="navbar">
      <NavLink exact to='/' className="nav-item" activeClassName='active-nav-item'>Saint-Petersburg</NavLink>
      <NavLink to={`/moscow/${moscowId}`} className="nav-item" activeClassName='active-nav-item'>Moscow</NavLink>
      <NavLink to={`/rostov-na-donu/${rostovId}`} className="nav-item" activeClassName='active-nav-item'>Rostov-na-Donu</NavLink>
    </nav>
  )
}
