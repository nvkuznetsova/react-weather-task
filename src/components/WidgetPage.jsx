import React, { Component } from 'react';

import { Loader } from './Loader';

export class WidgetPage extends Component {
  componentDidMount() {
    this.getForecastByCityId();
  }

  componentDidUpdate(prevProps) {
    if (this.props.match.params.id !== prevProps.match.params.id) {
      this.getForecastByCityId();
    }
  }

  getForecastByCityId() {
    const spbId = 498817;
    this.props.getWeatherForecast(this.props.match.params.id || spbId);
  }
  
  render() {
    if (this.props.isLoading === true) {
      return(
        <Loader />
      )
    } else {
      return (
        <div>
          <h1>Weather forecast for {this.props.weatherForecast.name}</h1>
        </div>
      )
    }
  }
}