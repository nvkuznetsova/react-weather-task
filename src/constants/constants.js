export const CITIES_IDS = Object.freeze({
  spb: 498817,
  moscow: 524901,
  rostov: 501175,
});

export const CITIES_NAMES = Object.freeze({
  spb: 'saint-petersburg',
  moscow: 'moscow',
  rostov: 'rostov-na-donu',
});