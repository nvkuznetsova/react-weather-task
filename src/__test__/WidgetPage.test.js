import React from 'react';

import { shallow } from 'enzyme';

import { WidgetPage } from '../components/WidgetPage';
import { CITIES_NAMES } from '../constants/constants';

describe('WidgetPage', () => {
  let component;
  const props = { 
    isLoading: true,
    getWeatherForecast: jest.fn(),
    match: {params: { cityName: CITIES_NAMES.spb}},
  };

  beforeAll(() => {
    component = shallow(<WidgetPage {...props}/>);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(component.instance()).toBeTruthy();
  });

  it('should call send weather request when component mount', () => {
    const cityId = 498817;
    const getForecastByCityId = jest.spyOn(component.instance(), 'getForecastByCityId');
    component.instance().componentDidMount();
    expect(getForecastByCityId).toHaveBeenCalled();
    expect(props.getWeatherForecast).toHaveBeenCalledWith(cityId);
  });

  it('should call send weather request when component update', () => {
    const prevProps = {
      ...props,
      match: {params: { cityName: CITIES_NAMES.moscow}}
    }
    const cityId = 498817;
    const getForecastByCityId = jest.spyOn(component.instance(), 'getForecastByCityId');
    component.instance().componentDidUpdate(prevProps);
    expect(getForecastByCityId).toHaveBeenCalled();
    expect(getForecastByCityId).toHaveBeenCalledTimes(1);
    expect(props.getWeatherForecast).toHaveBeenCalledWith(cityId);
  });

  describe('WidgetPage Redirect', () => {
    const redirectProps = {
      ...props,
      match: {params: { cityName: 'lalala'}},
    }

    beforeEach(() => {
      component = shallow(<WidgetPage {...redirectProps}/>);
    });

    it('should match redirect shapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('should change state and props when redirect', () => {
      expect(component.state().redirect).toBe(true);
      expect(component.props().to).toBe('page-not-found');
    });

    it('should redirect if incorrect URL', () => {
      expect(component.find('[data-marker="loader"]')).toHaveLength(0);
      expect(component.find('[data-marker="error"]')).toHaveLength(0);
      expect(component.find('[data-marker="weather-widget"]')).toHaveLength(0);
    });
  });

  describe('WidgetPage Loader', () => {
    const loaderProps = {...props};

    beforeEach(() => {
      component = shallow(<WidgetPage {...loaderProps}/>);
    });

    it('should match loading shapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('should display loader', () => {
      expect(component.find('[data-marker="loader"]')).toHaveLength(1);
      expect(component.find('[data-marker="error"]')).toHaveLength(0);
      expect(component.find('[data-marker="weather-widget"]')).toHaveLength(0);
    });
  });

  describe('WidgetPage Error', () => {
    const errorProps = {
      ...props,
      isLoading: false,
      error: { message: 'error' },
    };

    beforeEach(() => {
      component = shallow(<WidgetPage {...errorProps}/>);
    });

    it('should match error snapshot', ()=> {
      expect(component).toMatchSnapshot();
    });

    it('should display error', () => {
      expect(component.find('[data-marker="loader"]')).toHaveLength(0);
      expect(component.find('[data-marker="error"]')).toHaveLength(1);
      expect(component.find('[data-marker="weather-widget"]')).toHaveLength(0);
    });
  });

  describe('WidgetPage Widget', () => {
    const widgetProps = { 
      ...props,
      isLoading: false,
      error: { message: '' },
    };
    
    beforeEach(() => {
      component = shallow(<WidgetPage {...widgetProps}/>);
    });

    it('should match widget snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('should display widget', () => {
      expect(component.find('[data-marker="loader"]')).toHaveLength(0);
      expect(component.find('[data-marker="error"]')).toHaveLength(0);
      expect(component.find('[data-marker="weather-widget"]')).toHaveLength(1);
    });

    it('should contain header', () => {
      const title = component.instance().mainTitle.toUpperCase();
      expect(component.find('[data-marker="widget-title"]').text()).toBe(title);
    });
  });

});
