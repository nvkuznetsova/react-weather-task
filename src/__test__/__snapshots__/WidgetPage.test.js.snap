// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetPage WidgetPage Error should match error snapshot 1`] = `ShallowWrapper {}`;

exports[`WidgetPage WidgetPage Loader should match loading shapshot 1`] = `ShallowWrapper {}`;

exports[`WidgetPage WidgetPage Redirect should match redirect shapshot 1`] = `ShallowWrapper {}`;

exports[`WidgetPage WidgetPage Widget should match widget snapshot 1`] = `ShallowWrapper {}`;
